circuit/src/layered/circuit.rs

```diff
@@ -148,49 +149,97 @@ impl<C: GKRFieldConfig> Circuit<C> {
         rc.flatten::<Cfg>()
     }
 
-    pub fn load_non_simd_witness_file(&mut self, filename: &str) {
+    pub fn load_witness_allow_padding_testing_only(
+        &mut self,
+        filename: &str,
+        mpi_config: &MPIConfig,
+    ) {
         let file_bytes = fs::read(filename).unwrap();
-        self.load_witness_bytes(&file_bytes, true);
+        self.load_witness_bytes(&file_bytes, mpi_config, true, true);
     }
 
-    pub fn load_witness_file(&mut self, filename: &str) {
-        let file_bytes = fs::read(filename).unwrap();
-        self.load_witness_bytes(&file_bytes, false);
+    pub fn prover_load_witness_file(&mut self, filename: &str, mpi_config: &MPIConfig) {
+        let file_bytes = fs::read(filename)
+            .unwrap_or_else(|_| panic!("Failed to read witness file: {}", filename));
+        self.load_witness_bytes(&file_bytes, mpi_config, true, false);
+    }
+
+    pub fn verifier_load_witness_file(&mut self, filename: &str, mpi_config: &MPIConfig) {
+        let file_bytes = fs::read(filename)
+            .unwrap_or_else(|_| panic!("Failed to read witness file: {}", filename));
+        self.load_witness_bytes(&file_bytes, mpi_config, false, false);
     }
 
-    pub fn load_witness_bytes(&mut self, file_bytes: &[u8], allow_padding: bool) {
+    pub fn load_witness_bytes(
+        &mut self,
+        file_bytes: &[u8],
+        mpi_config: &MPIConfig,
+        is_prover: bool,
+        allow_padding_for_testing: bool, // TODO: Consider remove this
```

Comment:
the line on 158 is a bit hurting to me - hard to infer what 2 true in arguments meaning...

This actually starts to make me wonder what is the landscape and where we are at right now - I knew that we previously discussed that separating public input and witness, maybe this will help somewhat with `is_prover`, as verifier can load just the public input rather than reading the whole witness...

I guess the allow padding serves the same purpose of align statement-witness pairs up to pack size, now also with a factor of MPI size - but still pad with the first statement-witness pair, among all MPI parties?